### PR TITLE
Cygwinbrew :)

### PR DESCRIPTION
**Pull request by chrisjbreisch**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
This is an initial attempt to get Homebrew working for Cygwin.

```
$ uname -a
CYGWIN_NT-10.0-18362 Chris-PC 3.0.6-338.x86_64 2019-04-06 16:18 UTC x86_64 Cygwin

$ brew config
HOMEBREW_VERSION: 2.1.1-dirty
ORIGIN: https://github.com/Homebrew/brew
HEAD: b4f73e61649fcfc5aaa779c311e2514619ce01e7
Last commit: 4 days ago
Core tap ORIGIN: https://github.com/Homebrew/linuxbrew-core
Core tap HEAD: 45f8847270ea4574b0ca604a73bc15300a458950
Core tap last commit: 19 hours ago
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CACHE: /home/Chris/.cache/Homebrew
HOMEBREW_DEV_CMD_RUN: 1
HOMEBREW_DISPLAY: :0
HOMEBREW_LOGS: /home/Chris/.cache/Homebrew/Logs
HOMEBREW_NO_ANALYTICS_THIS_RUN: 1
HOMEBREW_NO_AUTO_UPDATE: 1
HOMEBREW_VISUAL: /usr/bin/vi
CPU: octa-core 64-bit skylake
Homebrew Ruby: 2.6.3 => /usr/bin/ruby.exe
Clang: N/A
Git: 2.17.0 => /bin/git
Curl: 7.59.0 => /usr/bin/curl
Kernel: CYGWIN_NT-10.0-18362 3.0.6-338.x86_64 x86_64 Cygwin
OS: Cygwin (3.0.6-338.x86_64)
Host glibc: 3.0
/usr/bin/gcc: 7.4.0
glibc: N/A
gcc: N/A
xorg: N/A

$ brew doctor
Your system is ready to brew.

$ type -a hello
bash: type: hello: not found

$ brew install hello
==> Downloading https://ftp.gnu.org/gnu/hello/hello-2.10.tar.gz
######################################################################## 100.0%
==> ./configure --disable-silent-rules --prefix=/home/linuxbrew/.linuxbrew/Cellar/hello/2.10
==> make install
�  /home/linuxbrew/.linuxbrew/Cellar/hello/2.10: 53 files, 400.8KB, built in 11 minutes 8 seconds

$ hello
Hello, world!
```

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/cleaner.rb [hunk 1]
```diff
@@ -2,6 +2,10 @@ class Cleaner
   private
 
   def executable_path?(path)
-    path.elf? || path.text_executable?
+    if OS.cygwin?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/cleaner.rb]: Changes such like should should be in an `extend/os/cygwin` layer.

Comment (chrisjbreisch) [reply Library/Homebrew/extend/os/linux/cleaner.rb]: Yes, I started down that route, but had issues. I'll see if I can work them out.

**Comment by chrisjbreisch** [comment]
One issue that I had to deal with is that the current Cygwin Ruby release is 2.3.6, and your minimum is 2.3.7. So, I downloaded, built and installed 2.6.3 (`./configure --prefix=/usr, make, make check, make install`). Without that, Homebrew tried to install a portable Linux Ruby bottle. That obviously didn't work well. I don't know if I should make a Cygwin bottle. If so, would need to change some code there so that it finds the correct bottle also.

**Comment by MikeMcQuaid** [comment]
@chrisjbreisch FYI we're unlikely to build/demand a Cygwin bottle due to it being pretty niche. A suggestion for now would be to hack the version check to allow it to work on any >=2.3 (which we would maybe consider including standalone) on non-macOS.

File: Library/Homebrew/extend/os/cygwin/cleaner.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class Cleaner
+  private
+
+  def executable_path?(path)
+    path.extname == ".exe" || path.text_executable?
```
Comment (chrisjbreisch) [inline Library/Homebrew/extend/os/cygwin/cleaner.rb]: I worked on an extended path for PE format executables, like the Linux one for ELF. I know the magic bytes. There's an "MZ" at byte 0x0 and a "PE" at byte 0x80, but for some reason my code didn't work. If you think this needs that, I can work on it.

**Comment by chrisjbreisch** [comment]
@MikeMcQuaid That works by me. I didn't know if you had a specific reason for setting the minimum to 2.3.7, so I chose not to mess with that.

**Comment by chrisjbreisch** [comment]
Sadly, the currently committed code no longer works in Cygwin. The issue reported at the end of this thread (https://github.com/Homebrew/linuxbrew-core/issues/3880) is not a fork one at all. It is due to the fact that Cygwin can not share open file descriptors with another process. I had put in a hack, but it broke the rest of Homebrew a little bit. The test in `formula_installer_spec.rb` starting at line 137 failed. 

I'll see what else I can come up with. For now, this PR is on hold.

**Comment by chrisjbreisch** [comment]
I have a working Homebrew in Cygwin once more. I can build `hello`, and I can run all tests in `utils/fork` and `formula_installer`. I haven't tried other tests in Cygwin yet. They'd take forever, and probably quite a few would fail anyway.

I'm not ready to push the new changes yet. I rewrote `safe_fork` to use a `TCPSocket` instead of a `UNIXSocket`. I figure that even if you were willing to make this change, that it would be a Cygwin only thing, so I need to make the changes necessary for both versions of `safe_fork` to be available. That shouldn't be too hard. One line of code in `build.rb` and `postinstall.rb` and then the specific versions of `safe_fork`.

**Comment by chrisjbreisch** [comment]
This is getting closer. It now should work on Cygwin, MacOS, and Linux. I've tested all three locally. However, there are still some issues on Cygwin.

I attempted a `brew install gcc` and had issues building `gmp`. One of them was that it was configuring with `--build=core2-linux-gnu`. `core2` seems a little odd to me. I would expect `x86_64`, but I guess `core2` is okay. `linux-gnu` is completely wrong, however. It should be `unknown-cygwin` or `pc-cygwin` or even just `cygwin`. I can't figure out where this string is built, so I've not been able to fix that yet.

File: Library/Homebrew/extend/os/linux/install.rb [hunk 3]
```diff
@@ -13,6 +13,8 @@ module Install
     ].freeze
 
     def symlink_ld_so
+      return if ENV["HOMEBREW_SYSTEM"] == "Cygwin"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/install.rb]: This should be in `extend/os/cygwin` instead.

File: Library/Homebrew/extend/os/linux/system_config.rb [hunk 4]
```diff
@@ -10,6 +10,9 @@ def host_os_version
         "#{description} (#{codename})"
       elsif (redhat_release = Pathname.new("/etc/redhat-release")).readable?
         redhat_release.read.chomp
+      elsif OS.cygwin?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/system_config.rb]: This should be in `extend/os/cygwin` instead.

File: Library/Homebrew/os.rb [hunk 5]
```diff
@@ -8,7 +8,14 @@ def self.mac?
   def self.linux?
     return false if ENV["HOMEBREW_TEST_GENERIC_OS"]
 
-    RbConfig::CONFIG["host_os"].include? "linux"
+    host_os = RbConfig::CONFIG["host_os"]
+    host_os.include?("linux") || host_os == "cygwin"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os.rb]: Cygwin should not be treated as Linux.

**Comment by MikeMcQuaid** [comment]
> we'd consider including it in parts later if it gets traction.

We'd also need some method of running Cygwin in CI if this ever happens.

File: Library/Homebrew/extend/os/linux/install.rb (hunk 3 above)
Comment (chrisjbreisch) [reply Library/Homebrew/extend/os/linux/install.rb]: Yes, my initial attempts were based on the information from this old request: https://github.com/Homebrew/linuxbrew-core/issues/3880. But have fixed that now.

File: Library/Homebrew/extend/os/linux/system_config.rb (hunk 4 above)
Comment (chrisjbreisch) [reply Library/Homebrew/extend/os/linux/system_config.rb]: Fixed

**Comment by chrisjbreisch** [comment]
> I think this changeset is too big to be included in Homebrew/brew unfortunately. I'd suggest running your own Cygwinbrew fork, adding it to https://github.com/Homebrew/brew/blob/master/docs/Interesting-Taps-and-Forks.md and we'd consider including it in parts later if it gets traction.

And you thought it was big before. :)

It now passes all style and unit tests in MacOS, Linux, & Cygwin. Cygwin takes a long time though. About an hour and a half on my machine.

**Comment by chrisjbreisch** [comment]
> > we'd consider including it in parts later if it gets traction.
> 
> We'd also need some method of running Cygwin in CI if this ever happens.

Yeah, that might be tricky. There's no CL installer for Cygwin. You'd probably have to install it somewhere and tar it up, then deploy the tarball to your build server.  In fact, you might want to put that in source control, so that it can be updated. Hmmm. Have to give that some thought. Awfully huge for source control, and it's binaries, and, and, and...blegh.

File: Library/Homebrew/os/mac/universal_server.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+require "extend/ENV"
+require "fcntl"
+require "socket"
+require "os/universal_server"
+
+class UniversalServer
+  def self.open
+    Dir.mktmpdir("homebrew", HOMEBREW_TEMP) do |tmpdir|
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/universal_server.rb]: Given this is identical on macOS and Linux it should just live in `Library/Homebrew/universal_server.rb` (with a better name like `Server`)

File: Library/Homebrew/os/mac/universal_socket.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require "extend/ENV"
+require "socket"
+
+class UniversalSocket
+  def self.open
+    socket = UNIXSocket.open(ENV["HOMEBREW_ERROR_PIPE"], &:recv_io)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/universal_socket.rb]: Given this is identical on macOS and Linux it should just live in `Library/Homebrew/universal_server.rb` (with a better name like just `Socket`)

File: Library/Homebrew/utils/os/mac/fork.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Utils
+  def self.fork_child_initialize(server)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/os/mac/fork.rb]: Given this is identical on macOS and Linux it should be in just `utils/fork.rb`

**Comment by chrisjbreisch** [comment]
> Just an FYI that this definitely won't be merged until we have a CI solution and non-trivial numbers of users (and even then given the test times you've mentioned I don't see it being terribly likely there). I'd strongly suggest you maintain your own repository and note it in the documentation as mentioned. We'd consider accepting some of these changes e.g. the `fork`, `socket`, `server`, minimum Ruby version stuff in a new PR.

Thanks, @MikeMcQuaid . I do appreciate your help and consideration. I'll continue to address your issues, and I may have another question or two.  In fact, I know I have at least one, but I'm away from my dev box at the moment and I need to look at the source.

**Pull request closed without merging**